src/Juvix/Backends/Michelson/Optimisation.hs

```diff
@@ -4,46 +4,53 @@
 module Juvix.Backends.Michelson.Optimisation where
 
 import Juvix.Library
+import Juvix.Backends.Michelson.Compilation.Types
 import Michelson.Typed
 
 {-  This is a simple optimization strategy which replaces sequences of Michelson instructions with equivalent sequences of fewer instructions.
-    Runs optimization passes until no further optimizations are found, up to a maximum number of passes.
+
     At the moment nontrivial programs are unlikely to compile to the smallest equivalent Michelson instruction sequence, but little time has been spent on optimization so far - a high degree should be possible; the Haskell typesystem provides very strong guarantees.
     A more interesting / potentially more effective strategy might be to search the space of equivalent Michelson programs, which at small program sizes using bounded heuristic search should be computationally feasible
       - then choose the one with the fewest instructions (or based on some other gas-estimation preference function).
     This optimization function is typed in the Instr GADT, so it cannot produce invalid output Michelson. However, the typesystem does not enforce computation correctness; that would require dependent types. -}
 
-optimise ∷ ∀ a b m. (Monad m) ⇒ Instr a b → m (Instr a b)
-optimise expr = do
-  let tellReturn ret = return ret -- tell [optimised (SomeExpr expr) (SomeExpr ret)] >> return ret
-      inner e = do
+optimise ∷ ∀ a b m.
+  (HasWriter "compilationLog" [CompilationLog] m) ⇒
+  Instr a b → m (Instr a b)
+optimise instr = do
+  let inner e = do
         one ← optimise' e
         two ← optimise' one
-        if one == two then return two else inner two
-  inner expr
+        if one == two then pure two else inner two
+  ret <- inner instr
+  tell @"compilationLog" [Optimised (SomeInstr instr) (SomeInstr ret)]
+  pure ret
 
-optimise' ∷ ∀ a b m. (Monad m) ⇒ Instr a b → m (Instr a b)
+optimise' ∷ ∀ a b m.
+  (Monad m) ⇒
+  Instr a b →
+  m (Instr a b)
 optimise' expr =
   case expr of
-    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF x y)
-    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_LEFT x y)
-    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_CONS x y)
-    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_NONE x y)
+    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF x y)
+    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_LEFT x y)
+    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_CONS x y)
+    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_NONE x y)
     (DIP e) → DIP |<< optimise' e
+#if defined(OPTIMISE)
```

Comment:
We could probably replace these CPP flags with some logic

Reply:
The goal is to reduce compilation time, so I don't think so... we could switch to the untyped form.

Reply:
I might have to do that anyways to make some optimisations which are semantically correct (mostly with `LAMBDA / EXEC`) but don't typecheck work...

Reply:
I will do this in a follow-up PR, I expect - ref https://github.com/cryptiumlabs/juvix/issues/168